crypto/bindings/srs.ts

```diff
@@ -0,0 +1,215 @@
+
+let cache: Cache | undefined;
+
+function setSrsCache(c: Cache) {
+  cache = c;
+}
+function unsetSrsCache() {
+  cache = undefined;
+}
+
+const srsVersion = 1;
+
+function cacheHeaderLagrange(f: 'fp' | 'fq', domainSize: number): CacheHeader {
+  let id = `lagrange-basis-${f}-${domainSize}`;
+  return withVersion(
+    {
+      kind: 'lagrange-basis',
+      persistentId: id,
+      uniqueId: id,
+      dataType: 'string',
+    },
+    srsVersion
+  );
+}
+function cacheHeaderSrs(f: 'fp' | 'fq', domainSize: number): CacheHeader {
+  let id = `srs-${f}-${domainSize}`;
+  return withVersion(
+    {
+      kind: 'srs',
+      persistentId: id,
+      uniqueId: id,
+      dataType: 'string',
+    },
+    srsVersion
+  );
+}
+
+function srs(wasm: Wasm, conversion: RustConversion) {
+  return {
+    fp: srsPerField('fp', wasm, conversion),
+    fq: srsPerField('fq', wasm, conversion),
+  };
+}
+
+function srsPerField(f: 'fp' | 'fq', wasm: Wasm, conversion: RustConversion) {
+  // note: these functions are properly typed, thanks to TS template literal types
+  let createSrs = wasm[`caml_${f}_srs_create_parallel`];
+  let getSrs = wasm[`caml_${f}_srs_get`];
+  let setSrs = wasm[`caml_${f}_srs_set`];
+
+  let maybeLagrangeCommitment = wasm[`caml_${f}_srs_maybe_lagrange_commitment`];
+  let lagrangeCommitment = wasm[`caml_${f}_srs_lagrange_commitment`];
+  let setLagrangeBasis = wasm[`caml_${f}_srs_set_lagrange_basis`];
+  let getLagrangeBasis = wasm[`caml_${f}_srs_get_lagrange_basis`];
+
+  return {
+    /**
+     * returns existing stored SRS or falls back to creating a new one
+     */
+    create(size: number): WasmSrs {
+      let srs = srsStore[f][size] as WasmSrs | undefined;
+
+      if (srs === undefined) {
+        if (cache === undefined) {
+          // if there is no cache, create SRS in memory
+          srs = createSrs(size);
+        } else {
+          let header = cacheHeaderSrs(f, size);
+
+          // try to read SRS from cache / recompute and write if not found
+          try {
+            let bytes = cache.read(header);
+
+            // TODO: this takes a bit to long, about 300ms for 2^16
+            // `pointsToRust` is the clear bottleneck
+            let jsonSrs: OrInfinityJson[] = JSON.parse(
+              new TextDecoder().decode(bytes)
+            );
+            let mlSrs = MlArray.mapTo(jsonSrs, OrInfinity.fromJSON);
+            let wasmSrs = conversion[f].pointsToRust(mlSrs);
+
+            srs = setSrs(wasmSrs);
+          } catch (e) {
+            // not in cache
+            srs = createSrs(size);
+
+            if (cache.canWrite) {
+              let wasmSrs = getSrs(srs);
+
+              let mlSrs = conversion[f].pointsFromRust(wasmSrs);
+              let jsonSrs = MlArray.mapFrom(mlSrs, OrInfinity.toJSON);
+              let bytes = new TextEncoder().encode(JSON.stringify(jsonSrs));
+
+              try {
+                cache.write(header, bytes);
+              } catch (e) {}
```

Comment:
we don't log the error?

Reply:
I wrote this before the cache had a `canWrite` property. so I had to swallow the error here to allow dummy / non-writing caches without spamming the console. However, now that there is `canWrite`, I will actually log the error here